### ISSUE TITLE
no-invalid-regexp: Make allowConstructorFlags option case sensitive

### ISSUE THREAD

**Issue by trosos** [eslint/eslint#16574]
Hi,

The `allowConstructorFlags` option of the `no-invalid-regexp` rule is case insensitive.
I believe that the option should be treated as case sensitive (because regexp flags are case sensitive).

### What rule do you want to change?

no-invalid-regexp

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
const ESLint = require("eslint");
console.log(new ESLint.Linter().verify(
  'RegExp(".", "A");',
  { rules: { "no-invalid-regexp": [1,
    { allowConstructorFlags: ["a"] }
  ] } }
));
```

### What does the rule currently do for this code?

The rule treats the option as case insensitive, and reports no issues.

### What will the rule do after it's changed?

The rule should IMO report invalid flag.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
I'm not sure why this option was made to be case-insensitive. There's no mention of this in https://github.com/eslint/eslint/pull/5352 or https://github.com/eslint/eslint/issues/5249.

By looking at the changes from https://github.com/eslint/eslint/pull/5352, it seems intentional as the regex that processes the option has the `i` flag, and there was this test: `{ code: "new RegExp('.', 'u')", options: [{ allowConstructorFlags: ["U"] }]}`. The test was later accidentally removed in https://github.com/eslint/eslint/commit/672deb057a14a7acad8c669189870009f1edb8a6.

I agree that this option should be case-sensitive, so that `allowConstructorFlags: ["a"]` does not allow `RegExp(".", "A")`, and that `allowConstructorFlags: ["A"]` does not allow `RegExp(".", "a")`. This change might be problematic though, as it might require some users to change their eslint configs and could be said to fall under "Major release > Part of the public API is removed or changed in an incompatible way > Rule schemas" in our [semver policy](https://github.com/eslint/eslint#semantic-versioning-policy).

Since this was apparently intentional, I'd like more input from the team on whether we should make this change (any ideas why this option is case-insensitive?) and whether we can treat it as a semver-minor bug fix.

**Comment by snitin315** [comment]
I'm 👍🏻 on this proposal as I also feel it should be case-sensitive. This would be a breaking change definitely so it should fall under the next major release.

**Comment by snitin315** [comment]
Marking this as accepted and adding it to the v9 board for tracking.

**Comment by mdjermanovic** [comment]
In `no-constructor-return`, instead of [`schema: {}`](https://github.com/eslint/eslint/blob/24206c49a138d4390f815ae122ee12f564bc604b/lib/rules/no-constructor-return.js#L23) it should be `schema: []`.

For example, `"no-constructor-return": ["error", "foo"]` should cause a validation error.

**Comment by JoshuaKGoldberg** [comment]
#17533 is ready for review. Marked as draft for now as it's a breaking change.
